compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1703,13 +1704,32 @@ fn add_post_link_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor
 /// that are necessary for the linking. They are only present in symbol table but not actually
 /// used in any sections, so the linker will therefore pick relevant rlibs for linking, but
 /// unused `#[no_mangle]` or `#[used]` can still be discard by GC sections.
+///
+/// There's a few internal crates in the standard library (aka libcore and
+/// libstd) which actually have a circular dependence upon one another. This
+/// currently arises through "weak lang items" where libcore requires things
+/// like `rust_begin_unwind` but libstd ends up defining it. To get this
+/// circular dependence to work correctly we declare some of this things
+/// in `symbols.o` file.
 fn add_linked_symbol_object(
     cmd: &mut dyn Linker,
     sess: &Session,
+    info: &CrateInfo,
     tmpdir: &Path,
     symbols: &[(String, SymbolExportKind)],
 ) {
-    if symbols.is_empty() {
+    let weak_symbols = (&info.used_crates)
+        .into_iter()
+        .filter_map(|cnum| info.missing_lang_items.get(&cnum))
```

Comment:
What we need here:
- Assumption 1: a weak lang item can be *required* by some crate and can be *provided* by some crate (typically higher in the dependency graph).
- Assumption 2: we cannot put nonexistent symbols to symbols.o, it may result in linking errors.
- Assumption 3: we add symbols to keep them alive, so linker doesn't throw away 1) their objects from static libraries that *provide* them and 2) maybe even whole dynamic libraries that *provide* then due to `--as-needed`, but that's pretty unrealistic.

- Conclusion: so we only need to add symbols that are *provided*, whether the symbol is actually *required* is not that important here, there are only 4 weak lang items right now so there are no performance issues.
- Possible cases:
	- The item is *provided* by the current crate.
		- Very well, nothing needs to be done, we don't even need to check our crate, just treat it as not *provided* and do not add it to symbols.o.
	- The item is *provided* by a static rlib dependency that is not a "through dylib" dependency (`Linkage::Static`, not `Linkage::IncludedFromDylib`).
		- Add it to symbols.o.
	- The item is *provided* by a dynamic dylib dependency (`Linkage::Dynamic`)
		- ??? Alternative 1: add it to symbols.o unless it breaks something ???
		- The case of a dylib *providing* lang items (typically libstd) being unnecessary and thrown away due to `--as-needed`, and library *requiring* them (typically libcore) being kept seems entirely fantastic though, so...
		- ??? Alternative 2: do nothing until someone provides a use case and reproduction ???
	- The item is *provided* by a static rlib dependency that is a "through dylib" dependency (`Linkage::IncludedFromDylib`).
		- Should be treated identically to the `Linkage::Dynamic` case above.

cc @nbdd0121 @bjorn3 for sanity checking

Reply:
> Assumption 1: a weak lang item can be required by some crate and can be provided by some crate (typically higher in the dependency graph).

True

> Assumption 2: we cannot put nonexistent symbols to symbols.o, it may result in linking errors.

Pretty sure that is true. Although `--gc-sections` may mask the errors.

> Assumption 3: we add symbols to keep them alive, so linker doesn't throw away 1) their objects from static libraries that provide them and 2) maybe even whole dynamic libraries that provide then due to --as-needed, but that's pretty unrealistic.

This is true for all symbols in `symbols.o` before this PR. As for weak lang items, ld.bfd and ld.gold and other traditional linkers only resolve symbols in later object files. They don't resolve them in object files included earlier in the commandline arguments. This is why `--start-group` and `--end-group` were needed. I think including them in `symbols.o` also works. Lld and mold happily resolve symbols anywhere though.

> Conclusion: so we only need to add symbols that are provided, whether the symbol is actually required is not that important here, there are only 4 weak lang items right now so there are no performance issues.

Adding symbols that are not required may cause more functions than required to be included in the binary. No other side effects though. Note that `missing_lang_items` only includes lang items that are actually necessary.